frontend/src/pages/RecurringRunDetails.test.tsx

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import RecurringRunDetails from './RecurringRunDetails';
+import TestUtils from '../TestUtils';
+import produce from 'immer';
+import { ApiJob, ApiResourceType } from '../apis/job';
+import { Apis } from '../lib/Apis';
+import { PageProps } from './Page';
+import { RouteParams, RoutePage } from '../components/Router';
+import { ToolbarActionConfig } from '../components/Toolbar';
+import { shallow } from 'enzyme';
+
+class TestRecurringRunDetails extends RecurringRunDetails {
+  public async _deleteDialogClosed(confirmed: boolean): Promise<void> {
+    super._deleteDialogClosed(confirmed);
+  }
+}
+
+describe('RecurringRunDetails', () => {
+  const updateBannerSpy = jest.fn();
+  const updateDialogSpy = jest.fn();
+  const updateSnackbarSpy = jest.fn();
+  const updateToolbarSpy = jest.fn();
+  const historyPushSpy = jest.fn();
+  const getJobSpy = jest.spyOn(Apis.jobServiceApi, 'getJob');
+  const deleteJobSpy = jest.spyOn(Apis.jobServiceApi, 'deleteJob');
+  const enableJobSpy = jest.spyOn(Apis.jobServiceApi, 'enableJob');
+  const disableJobSpy = jest.spyOn(Apis.jobServiceApi, 'disableJob');
+  const getExperimentSpy = jest.spyOn(Apis.experimentServiceApi, 'getExperiment');
+
+  let fullTestJob: ApiJob = {};
+
+  function generateProps(): PageProps {
+    return {
+      history: { push: historyPushSpy } as any,
+      location: '' as any,
+      match: { params: { [RouteParams.runId]: fullTestJob.id }, isExact: true, path: '', url: '' },
+      toolbarProps: RecurringRunDetails.prototype.getInitialToolbarState(),
+      updateBanner: updateBannerSpy,
+      updateDialog: updateDialogSpy,
+      updateSnackbar: updateSnackbarSpy,
+      updateToolbar: updateToolbarSpy,
+    };
+  }
+
+  beforeEach(() => {
+    fullTestJob = {
+      created_at: new Date(2018, 8, 5, 4, 3, 2),
+      description: 'test job description',
+      enabled: true,
+      id: 'test-job-id',
+      max_concurrency: '50',
+      name: 'test job',
+      pipeline_spec: {
+        parameters: [{ name: 'param1', value: 'value1' }],
+        pipeline_id: 'some-pipeline-id',
+      },
+      trigger: {
+        periodic_schedule: {
+          end_time: new Date(2018, 10, 9, 8, 7, 6),
+          interval_second: '3600',
+          start_time: new Date(2018, 9, 8, 7, 6),
+        }
+      },
+    } as ApiJob;
+
+    historyPushSpy.mockClear();
```

Comment:
I guess it depends on what your intent is with the `mockImplementation()` calls. If you're just trying to remove whatever other mock implementation there was, you should use `mockReset`, if you actually want there to be an empty implementation, then calling `clear` + `implement` is right.

In the cases where you're calling `mockClear` without `mockImplementation` though, it seems like `reset` should be preferable because that way you make sure you remove whatever the previous mocked implementation was